src/runtime-rs/crates/hypervisor/src/qemu/inner.rs

```diff
@@ -167,10 +167,19 @@ impl QemuInner {
     pub(crate) async fn stop_vm(&mut self) -> Result<()> {
         info!(sl!(), "Stopping QEMU VM");
         if let Some(ref mut qemu_process) = &mut self.qemu_process {
-            info!(sl!(), "QemuInner::stop_vm(): kill()'ing qemu");
-            qemu_process.kill().await.map_err(anyhow::Error::from)
+            let is_qemu_running = qemu_process.id().is_some();
+            if is_qemu_running {
+                info!(sl!(), "QemuInner::stop_vm(): kill()'ing qemu");
+                qemu_process.kill().await.map_err(anyhow::Error::from)
+            } else {
+                info!(
+                    sl!(),
+                    "QemuInner::stop_vm(): qemu process isn't running (likely stopped already)"
```

Comment:
It is really unfortunate for that condition to be handled here but it is certainly the best option for now.
If you happen to re-post, please add some `// FIXME` to make it clear that this is a workaround that will need to be revisited (and not blindly copied into a future hypervisor driver).

Reply:
Note that I'm unfortunately _not_ sure that this is a work-around. :-( It could be that the shim v2 protocol just doesn't offer a better opportunity than the WaitProcess handler to stop the sandbox.

Reply:
Not being sure about how this is expected to work is even a better reason for a comment :wink:. At this stage we should know if shim v2 allows that and if it doesn't we should put a comment to explain the flow.

Reply:
The purpose of this PR is simply to make the qemu driver's behaviour consistent with other drivers whose `stop_vm()` implementations are (silently) idempotent.

I don't feel comfortable to comment on the call pattern of the function since I don't have the necessary understanding, just suspicions.  I raised a question above to clear this up but unless someone who does know chimes in we simply have to accept that we currently don't know what the right solution is.  I will continue investigating the issue but that's far beyond the scope of this PR.